### ISSUE TITLE
Add support for HomeKit displays?

### ISSUE THREAD

**Issue by Tediore** [jlusiardi/homekit_python#171]
My Vizio TV was just updated with HomeKit support. The TV pairs with HASS, but no entity appears. I don't need this feature, but others might want it. I can help with testing if this is something that should be added to the platform.

**Comment by jlusiardi** [comment]
Hey,

Thanks for your oder! Great to hear you are already able to pair with HASS. 
I appreciate your offer and will send you a list of things you might test for me, would that be ok?

Joachim

**Comment by Tediore** [comment]
Yes, sounds good.

**Comment by jlusiardi** [comment]
Hey @Tediore,

could you please do the following for a start:

1. unpair the tv from HASS
2. clone this repository
3. checkout our working branch `git checkout fix_171_homekit_displays`
3. perform a `python3 -m homekit.init_controller_storage -f tv.json`
4. perform a `python3 -m homekit.discover` and post the output related to the TV
5. perform a pairing with `python3 -m homekit.pair -d ${DEVICEID} -p RED-AC-TED -f tv.json -a TV  --log DEBUG` (this should work, if not, pls post the output)
6. perform `python3 -m homekit.get_accessories -f tv.json -a TV -o compact --log DEBUG` and `python3 -m homekit.get_accessories -f tv.json -a TV -o json ` and give me the output

From that we can see the services and characteristics of the tv.

**Comment by Tediore** [comment]
I've cloned the repo, but when I run `git checkout fix_171_homekit_displays` I get the following error:
`fatal: Not a git repository (or any of the parent directories): .git`

What am I doing wrong?

**Comment by Tediore** [comment]
Here's everything I ran in the terminal:
`mattjwydra@penguin:~$ git clone https://github.com/jlusiardi/homekit_python.git
Cloning into 'homekit_python'...
remote: Enumerating objects: 116, done.
remote: Counting objects: 100% (116/116), done.
remote: Compressing objects: 100% (83/83), done.
remote: Total 3420 (delta 62), reused 60 (delta 32), pack-reused 3304
Receiving objects: 100% (3420/3420), 1.21 MiB | 0 bytes/s, done.
Resolving deltas: 100% (2432/2432), done.
mattjwydra@penguin:~$ git checkout fix_171_homekit_displays
fatal: Not a git repository (or any of the parent directories): .git`

**Comment by Tediore** [comment]
Here's the full terminal output:
`tediore@penguin:~$ git clone https://github.com/jlusiardi/homekit_python.git
Cloning into 'homekit_python'...
remote: Enumerating objects: 116, done.
remote: Counting objects: 100% (116/116), done.
remote: Compressing objects: 100% (83/83), done.
remote: Total 3420 (delta 62), reused 60 (delta 32), pack-reused 3304
Receiving objects: 100% (3420/3420), 1.21 MiB | 0 bytes/s, done.
Resolving deltas: 100% (2432/2432), done.
tediore@penguin:~$ git checkout fix_171_homekit_displays
fatal: Not a git repository (or any of the parent directories): .git`

**Comment by jlusiardi** [comment]
Sorry, I didn't know your level of shell knowledge and setup. I added some steps that were missing.

Steps for preparation:
1. unpair the tv from HASS
2. clone this repository
3. change into cloned repository: `cd homekit_python` 
3. checkout our working branch: `git checkout fix_171_homekit_displays`
3. create a virtual environment: `python3 -m venv venv` (package under debian: `python3.7-venv`)
3. activate virtual environment: `source venv/bin/activate`
3. install python_homekit and dependencies into virtual environment: `pip install .[IP,BLE]`

Steps for testing:
1. make sure virtual environment is activate: `source venv/bin/activate`
3. initialize storage for controller's crypto data:  `PYTHONPATH=. python3 homekit/init_controller_storage.py -f tv.json`
4. I assume the TV is connected via IP (and not Bluetooth LE) so perform a `PYTHONPATH=. python3 homekit/discover.py` and post the output related to the TV
5. perform a pairing with `PYTHONPATH=. python3 homekit/pair.py -d ${DEVICEID} -p RED-AC-TED -f tv.json -a TV  --log DEBUG` (this should work since you were able to pair with HASS, if not, pls post the output)
6. perform `PYTHONPATH=. python3 homekit/get_accessories.py -f tv.json -a TV -o compact --log DEBUG` and `python3 -m homekit.get_accessories -f tv.json -a TV -o json ` and give me the output

From that we can see the services and characteristics of the tv.

**Comment by Tediore** [comment]
No problem. Sorry about that, I should have mentioned that I'm very much a beginner, but I'm willing to learn and help. I got up up to step 7 and received this error for each wheel it tried to build:
`error: invalid command 'bdist_wheel'`

[Terminal output from step 7](https://pastebin.com/amKKycAj)

**Comment by jlusiardi** [comment]
Hey,

can you tell me which distribution you have? 
On Debian the file `Python.h` could be from `libpython3.5-dev`. Can you try again after installing?

**Comment by Tediore** [comment]
Debian. Will do, sorry for the trouble.

**Comment by Tediore** [comment]
Figured it out, I was missing some dependencies. Running `PYTHONPATH=. python3 homekit/discover.py` doesn't result in any output in the terminal or in tv.json. As a result, I'm assuming I don't have the DEVICEID  and PIN that I need to do the next command.

**Comment by jlusiardi** [comment]
`PYTHONPATH=. python3 homekit/discover.py` should return something like 
```
Name: DemoAccessory._hap._tcp.local.
Url: http_impl://127.0.0.1:8080
Configuration number (c#): 12
Feature Flags (ff): No support for HAP Pairing (Flag: 0)
Device ID (id): 12:00:00:00:00:00
Model Name (md): DemoAccessory
Protocol Version (pv): 1.0
State Number (s#): 1
Status Flags (sf): Accessory has been paired. (Flag: 0)
Category Identifier (ci): Lightbulb (Id: 5)
```
And the interesting thing is the line with `Device ID (id): 12:00:00:00:00:00`. I guess, since the TV obviously has a display, the required pin code will be displayed there. Perhaps you must activate a special mode on the TV before?
Use `PYTHONPATH=. python3 homekit/pair.py -d ${DEVICEID}  -f tv.json -a TV --log DEBUG` and the python script should ask you to enter the pin.

**Comment by Tediore** [comment]
Yeah, unfortunately I've tried everything I can think of at this point. I confirmed on the TV that it's unpaired from HomeKit and it shows up in HASS discovery...so it's broadcasting that it's available to be paired but the terminal command isn't finding it for some reason.

**Comment by Jc2k** [comment]
So HASS and homekit_python both ultimately use the same zeroconf library, there shouldn't really be much that could stop it working. homekit_python only ignores zeroconf results that don't have a HomeKit model or configuration model, i think hass tries to continue if they are missing, but thats a massive (and obvious) spec violation so i dont think it would be that.

Do any other devices show up? (the discovery command should see paired devices too). Are you running this on the same machine as your home assistant install? Is it a VM? The most likely problem would be if you had a seperate VM for testing this and it was NATted rather than bridged. Are HASS and this Debian machine on the same network and specifically the same subnet? Both wifi/physical connection? Or different?

**Comment by Tediore** [comment]
Understood, good to know. 

No other devices show up, but the only device that I have paired with HK controller is my Ecobee. It's a Debian VM on my chromebook which is on the same network and subnet as all other devices. Different machine than my HASS setup. The chromebook and TV are connected over wifi.

**Comment by Jc2k** [comment]
I don’t really know the chromebook stack but given HASS can see it id assume it was something on the chromebook- can you ssh into the HASS box and run command there?

**Comment by jlusiardi** [comment]
Just a guess: 

> It's a Debian VM on my chromebook which is on the same network and subnet as all other devices. 

This means you are using bridged networking and not some kind of NAT for the VM? Can you ping the TV?

**Comment by Tediore** [comment]
@Jc2k I can try that. 
@jlusiardi Correct. I can ping the TV, and `nmap` shows all the devices on my network.

**Comment by jlusiardi** [comment]
nmapping and pinging the tv only means you can connect to it. can you post the IPs of the TV and the VM? And perhaps do a traceroute?

**Comment by Tediore** [comment]
Before I try anything else, I'll ssh into my Pi and run the commands from there. But yes, I'll also run a traceroute.
Sorry for the trouble, hopefully this isn't more of a pain than it's worth.

**Comment by Jc2k** [comment]
Another thing is that the HASS instance will have captured and cached much of the data we are after already. When the TV is paired with HASS, in the directory where your configuration.yml lives you will find a .storage folder, and inside that one of the files should be called something like `homekit_controller-entity-map`. That will have JSON for all of your paired devices (including your ecobee). So if you can't get the CLI to work we might be able to make do with that.

**Comment by Tediore** [comment]
Ugh I completely forgot about that...I've poked around in that file before too. Well, I'll take a look at everything tonight. Also, if it's relevant, I'm running Hassio on Raspbian.
It might also be interesting to note that I can no longer pair the TV to HASS since I paired and unpaired it the first time. HASS reports it discovered a new device, but when I click on _configure_ in the HASS integrations screen, the loading icon comes up, but the pairing dialog (to enter the HomeKit code) never appears. Also, when I click add new integration (or whatever it is) and click _homekit accessory_, it displays the same behavior and then returns to the previous screen. Might not be relevant to the issue at hand, but figured I'd mention it.

**Comment by Tediore** [comment]
So I SSH'd into my pi and tried running the commands. Running `pip install .[IP,BLE]` gave me the same warnings as before about missing dependencies, so I installed those. It still gives me a long error at the end, so I decided to run `pip install .[IP]` which worked except it said `failed to install homekit`. 

Either way, running the discovery command worked, but it doesn't detect the TV. It only detects all the other HomeKit compatible devices.

**Comment by jlusiardi** [comment]
First of all, is this still a beta version of the homekit support? Or the stable version? It could be that we are hunting a bug in VIZIO's implementation.

> HASS reports it discovered a new device

Is this the TV? If so, is there any information on screen or in the logs that look like a Device ID (12:34:56:78:90:12 for example)?

> paired and unpaired it the first time

Perhaps something failed. Maybe there is an option to reset the homekit stuff (without reseting the other settings)?

**Comment by Jc2k** [comment]
Related to:

> paired and unpaired it the first time

Did you unpair on the TV? Or from within HASS. HASS won't actually tell the device to unpair yet, so you have to unpair on the TV end.

That said, a paired TV should still show in the discover output, unless they don't advertise when paired for some reason?

**Comment by Tediore** [comment]
@jlusiardi: Yes, HASS is detecting the TV again. I can only assume it's a stable version since I never signed up for beta firmware for my TV, but I'll take a look at that. I don't see anything that looks like a device ID.

@Jc2k: I unpaired from within HASS and from the TV itself. I can pull up the pairing screen and HomeKit code on the TV.

**Comment by jlusiardi** [comment]
@tediore, so hassio can discover the TV from your Raspberry pi. That is good news. Can you try to execute the steps from above on the raspberrry? Perhaps we removed some functionality by accident.

**Comment by Tediore** [comment]
I tried running the commands on the pi and the discovery command didn't list the TV, so I unfortunately couldn't try the pairing command.

**Comment by jlusiardi** [comment]
ok, can you please run `git checkout 0.15.0`. This should go back to the current released state and then try running the commands again? 

(Sorry the reply took so long, been busy at work)

**Comment by Tediore** [comment]
Sorry for the incredibly late reply. I tried pairing with the TV again. Here's what discovery shows for it:
```
Name: Living room TV._hap._tcp.local.
Url: http_impl://10.0.0.50:54123
Configuration number (c#): 3996
Feature Flags (ff): No support for HAP Pairing (Flag: 8)
Device ID (id): redacted
Model Name (md): E55-E2
Protocol Version (pv): 1.1
State Number (s#): 1
Status Flags (sf): Accessory has not been paired with any controllers. (Flag: 1)
Category Identifier (ci): Unknown Category #31 (Id: 31)
```
When I try to pair it, the TV recognizes that it's paired to a controller, but homekit python doesn't appear to. Here's the output from the pairing process: https://paste.ubuntu.com/p/k6gnfhkb6N/

When I try to run `python3 homekit/get_accessories.py -f tv.json -a TV --log DEBUG`, it prints `"TV" is no known alias"`

**Comment by Jc2k** [comment]
This looks like a `cryptrography` library version error. Can you do `pip list | grep cryptography`. You need version 2.5 or higher.

**Comment by Tediore** [comment]
I have version 2.6.1 installed.

**Comment by Jc2k** [comment]
In my dev env I see this:

```
(venv) % python
Python 3.7.3 (default, Jun 19 2019, 07:38:49) 
[Clang 10.0.1 (clang-1001.0.46.4)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from cryptography.hazmat.serialization import Encoding
KeyboardInterrupt
>>> from cryptography.hazmat.primitives import serialization
>>> serialization.Encoding.Raw
<Encoding.Raw: 'Raw'>
>>> 
```

I tested with a fresh venv and `cryptography==2.6.1`, and i have the attribute that your env thinks is missing.

What is your host OS? I wonder if your openssl version is too old.

**Comment by Tediore** [comment]
I'm running Raspbian 10. I have OpenSSL 1.1.1c. `apt upgrade` tells me there's a newer version, so I'll update to that.

**Comment by Jc2k** [comment]
I think 1.1.1c should be new enough, and Raspbian 10 should be new enough too.

Can you run the commands i pasted in the shell above and see if you get the same result?

**Comment by Tediore** [comment]
Unfortunately I get the same error as before:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.7/enum.py", line 349, in __getattr__
    raise AttributeError(name) from None
AttributeError: Raw
```

**Comment by Jc2k** [comment]
What about

```
Python 3.7.3 (default, Jun 19 2019, 07:38:49) 
[Clang 10.0.1 (clang-1001.0.46.4)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import cryptography
>>> cryptography.__path__
['/Users/john/Projects/home-assistant/tedi/lib/python3.7/site-packages/cryptography']
>>> cryptography.__version__
'2.6.1'
```

**Comment by Tediore** [comment]
Well there it is. `'2.3.1'`.

**Comment by Jc2k** [comment]
Ok - it will be too hard to figure out how but you have some sort of contamination in how you are running the commands so the host (maybe) cryptography package is getting used but `pip` can see a different one. The host one is too old, hence the exception. Lets just start from the top with a fresh virtualenv.

```
$ cd ~
$ mkdir test-pairing
$ cd test-pairing/
$ python3 -m venv venv
$ ./venv/bin/pip install homekit
Collecting homekit
  Using cached https://files.pythonhosted.org/packages/20/a2/cd9de6add789c8778657a3d04a82bfce2879d104dfa09c26d8ea0d0435b8/homekit-0.15.0.tar.gz
...snip....
  Running setup.py install for homekit ... done
Successfully installed cffi-1.13.2 cryptography-2.8 ed25519-1.5 hkdf-0.0.3 homekit-0.15.0 pycparser-2.19 six-1.13.0
$ ./venv/bin/python -m homekit.discover
$ ./venv/bin/python -m homekit.init_controller_storage -f pairings.json
$ ./venv/bin/python -m homekit.pair -f pairings.json -d xx:xx:xx:xx:xx -a mytv
It should prompty you for a pin here, make sure you enter the dashes `xxx-xx-xxx`.
$ ./venv/bin/python -m homekit.get_accessories -f pairings.json -a mytv
```

I've tried to use the most explicit ways of running the commands. E.g. its quite common for `pip` or `pip3` to be a different one to what you expect because of the value of $PATH so some people run it with `python -m pip`, then you at least know its the pip that matches the python you are using. Likewise, the transition from python2 to python3 has introduced a lot of confusion around whether `python` is 2 or 3 and whether `pip` is 2 or 3. I've explicitly used `python3` to create a virtual environment in case the host os uses `python` as python2, but within the virtualenv i know `python` is python 3.

TL;DR if you run those commands as written i'd expect the cross contamination issue to go away and hopefully it will work and you'll pair with your TV.

**Comment by Tediore** [comment]
Okay, figured it out. Explanation after. 

Here's the output from `python3 homekit/get_accessories.py -f tv.json -a TV -o compact --log DEBUG`: https://paste.ubuntu.com/p/Zs668mgMxf/

And here's the output from `python3 -m homekit.get_accessories -f tv.json -a TV -o json`: https://paste.ubuntu.com/p/Qwn4j6CVVY/

So...one of my other venvs (`xbox-smartglass-rest`) that I created after I created the homekit one had an issue. I ran `which python` with the smartglass venv activated and it was showing a path outside the venv, so when I installed `xbox-smartglass-rest` in that one (which needs `cryptography==2.3.1`), it was installing `cryptography==2.3.1` outside the venv. I deleted and recreated that venv and created a fresh homekit one too. Sorry about that, didn't mean to turn this into a troubleshooting session with my system.

Thank you for your patience.

**Comment by jlusiardi** [comment]
Hey @Jc2k 

Thanks for figuring that out! 

Hey @Tediore 

Cool to see the progress! 

Awesome job by the both of you!

**Comment by jlusiardi** [comment]
So, since it was some virtual env with some problems, can we close this one here? 

@Tediore could you try to make a pull request with a filled https://github.com/jlusiardi/homekit_python/blob/master/doc/tested_devices/Homekit%20Device%20Test%20Template.ipynb?

**Comment by Jc2k** [comment]
Not entirely i guess - there are a bunch of characteristics in there that we don't understand. E.g. service D8:

```
            {
                "characteristics": [
                    {
                        "format": "string",
                        "iid": 14,
                        "perms": [
                            "pw",
                            "pr",
                            "ev"
                        ],
                        "type": "E3",
                        "value": "Living room TV"
                    },
                    {
                        "format": "uint8",
                        "iid": 15,
                        "maxValue": 1,
                        "minStep": 1,
                        "minValue": 0,
                        "perms": [
                            "pw",
                            "pr",
                            "ev"
                        ],
                        "type": "000000B0-0000-1000-8000-0026BB765291",
                        "value": 1
                    },
                    {
                        "format": "uint32",
                        "iid": 16,
                        "perms": [
                            "pw",
                            "pr",
                            "ev"
                        ],
                        "type": "000000E7-0000-1000-8000-0026BB765291",
                        "value": 7
                    },
                    {
                        "format": "uint8",
                        "iid": 17,
                        "maxValue": 1,
                        "minStep": 1,
                        "minValue": 0,
                        "perms": [
                            "pr",
                            "ev"
                        ],
                        "type": "E8",
                        "value": 1
                    },
                    {
                        "format": "uint8",
                        "iid": 18,
                        "maxValue": 1,
                        "minStep": 1,
                        "minValue": 0,
                        "perms": [
                            "pw",
                            "pr",
                            "ev"
                        ],
                        "type": "DD",
                        "value": 0
                    },
                    {
                        "format": "uint8",
                        "iid": 19,
                        "maxValue": 1,
                        "minStep": 1,
                        "minValue": 0,
                        "perms": [
                            "pw"
                        ],
                        "type": "DF"
                    },
                    {
                        "format": "int",
                        "iid": 20,
                        "maxValue": 100,
                        "minStep": 1,
                        "minValue": 0,
                        "perms": [
                            "pw",
                            "pr",
                            "ev"
                        ],
                        "type": "00000008-0000-1000-8000-0026BB765291",
                        "unit": "percentage",
                        "value": 50
                    },
                    {
                        "format": "uint8",
                        "iid": 21,
                        "maxValue": 13,
                        "minStep": 1,
                        "minValue": 0,
                        "perms": [
                            "pw",
                            "pr",
                            "ev"
                        ],
                        "type": "E2",
                        "value": 0
                    },
                    {
                        "format": "uint8",
                        "iid": 22,
                        "maxValue": 16,
                        "minStep": 1,
                        "minValue": 0,
                        "perms": [
                            "pw"
                        ],
                        "type": "E1"
                    },
                    {
                        "format": "uint8",
                        "iid": 23,
                        "maxValue": 3,
                        "minStep": 1,
                        "minValue": 0,
                        "perms": [
                            "pr",
                            "ev"
                        ],
                        "type": "E0",
                        "value": 3
                    },
                    {
                        "format": "tlv8",
                        "iid": 24,
                        "perms": [
                            "pr",
                            "ev"
                        ],
                        "type": "136",
                        "value": "AQQBAAAAAAABBAIAAAAAAAEEAwAAAAAAAQQEAAAAAAABBAUAAAAAAAEEBgAAAAAAAQQHAAAAAAABBAgAAAAAAAEECQAAAA=="
                    }
                ],
                "hidden": false,
                "iid": 13,
                "linked": [
                    31,
                    39,
                    47,
                    55,
                    63,
                    71,
                    78,
                    85,
                    92,
                    1,
                    8,
                    10,
                    25
                ],
                "primary": true,
                "type": "D8"
            },
```

There is data we can use in HAP-nodejs here - https://github.com/KhaosT/HAP-NodeJS/blob/master/src/lib/gen/HomeKit-TV.ts.

But I can't see TV's mentioned in the R2 docs or the open source HAP implementation from Apple.

**Comment by jlusiardi** [comment]
@Jc2k can you have a look for branch fix_171_include_service_and_characteristics_for_television where i try to add support for TVs?

**Comment by Jc2k** [comment]
Had a quick look earlier, looks good. Will see if it can parse the sample data when I get a minute.

**Comment by Tediore** [comment]
Need me to do anything?

**Comment by Jc2k** [comment]
@Tediore I'd quite like to see the output of `python3 homekit/get_accessories.py -f tv.json -a TV -o compact` on @jlusiardi 's  [fix_171_include_service_and_characteristics_for_television](https://github.com/jlusiardi/homekit_python/tree/fix_171_include_service_and_characteristics_for_television) branch if you are able to help. It should be able to decode more of the field names now.

**Comment by Tediore** [comment]
Sure thing, I can do that.

**Comment by Tediore** [comment]
Welp, I'm not sure this will be much help, but here's the output from that command after I checked out that branch: https://paste.ubuntu.com/p/PxhBsZytQg/

**Comment by jlusiardi** [comment]
Hi @Tediore,

I added some more fixes and am now able to start a demoTV (not usable with an iOS controller though). 

Using a file `demoTv.json` like 
```
{
  "accessory_pairing_id": "12:00:00:00:01:0C",
  "accessory_pin": "132-45-786",
  "c#": 3,
  "category": "Television",
  "host_ip": "127.0.0.1",
  "host_port": 51841,
  "name": "demoTV",
  "peers": {},
  "unsuccessful_tries": 0
}
``` 
gives me an output after pairing and executing `PYTHONPATH=. python3 homekit/get_accessories.py -f test_controller.json -a demo` like following:

```
1.2: >accessory-information<
  1.3:  (Identify) >identify< [pw]
  1.4: lusiardi.de (Manufacturer) >manufacturer< [pr]
  1.5: Demoserver (Model) >model< [pr]
  1.6: TestTV (Name) >name< [pr]
  1.7: 0001 (Serial Number) >serial-number< [pr]
  1.8: 0.1 (Firmware Revision) >firmware.revision< [pr]
1.9: >television<
  1.10: 0 (Active state (Inactive/Active)) >active< [pw,pr]
  1.11: 0 (Indicate target as selected in UI) >active-identifier< [pr]
  1.12: None (Configured Name) >configured-name< [pw,pr]
  1.13: 0 (Active state (Inactive/Active)) >sleep-discovery-mode< [pr]
  1.14: 0 (Brightness in percent) >brightness< [pr,pw,ev]
  1.15: 0 (Active state (Inactive/Active)) >closed-captions< [pr,pw]
  1.16: None () >display-order< [pr,pw]
  1.17: 0 (Current media state) >current-media-state< [pr,pw]
  1.18: 0 (Target media state) >target-media-state< [pr,pw]
  1.19: 0 (Picture Mode) >picture-mode< [pr,pw]
  1.20:  (Active state (Inactive/Active)) >picture-mode< [pw]
  1.21:  (Target media state) >remote-key< [pw]
```

Can you try again please?

**Comment by Tediore** [comment]
I pulled the latest changes from the `include_service...` branch and ran the command again. Unfortunately I still see a lot of "unknown characteristic": https://paste.ubuntu.com/p/C6HzF2K68N/

**Comment by jlusiardi** [comment]
sorry for the inconvenience. can you run 
```
python3 -m venv test_env
source test_env/bin/activate
pip3 install -r requirements.txt 
PYTHONPATH=. python3 homekit/get_accessories.py -f test_controller.json -a demoTV 
PYTHONPATH=. python3 homekit/get_accessories.py -f test_controller.json -a demoTV -o json
```
and post the output? I just want to make sure, no old version of the library is used. Thanks!

**Comment by Tediore** [comment]
So I ran that and installed some things I didn't have before and accidentally exited out of the terminal window...sorry about that. Nevertheless, I ran the command again that @Jc2k requested and unfortunately it returned the same output.

**Comment by jlusiardi** [comment]
Hey @Tediore,
no problem, short explanation what the commands do:

`python3 -m venv test_env` creates a virtual python environment. This is handy to have different setups for different tools (allowing different versions of the same library installed in each virtual env). needs to be done once.

`source test_env/bin/activate` activates this environment, meaning you use only the packages installed in there by fiddling around with some env vars. needs to be done every time you have a new shell.

`pip3 install -r requirements.txt` installs the requirements listed in the file (if you are in a virtual environment, theses packages are installed there and won't clutter your system).

if you run python commands now, they'll use the packages from the virtual env.

So please go back to the directory you cloned the master into end execute:
```
source test_env/bin/activate
PYTHONPATH=. python3 homekit/get_accessories.py -f test_controller.json -a demoTV 
PYTHONPATH=. python3 homekit/get_accessories.py -f test_controller.json -a demoTV -o json
```

**Comment by Tediore** [comment]
Oh sorry, I explained that poorly. I'm familiar with those commands (I usually spin up a new venv if I install a new command line tool to avoid dependency issues). Whenever I've been running the discovery, pairing, etc. commands you guys have been giving me I've had the venv activated and have been in the homekit_python directory. I thought for some reason you wanted to see the terminal output from when I ran `pip3 install -r requirements.txt`.

Anyway, running `PYTHONPATH=. python3 homekit/get_accessories.py -f test_controller.json -a demoTV` returns `Could not open "test_controller.json" because it does not exist`
If I initialize the controller storage with that file name, running the command returns `"demoTV" is no known alias.`

**Comment by jlusiardi** [comment]
Ok sorry,

Do you have the pairing data from your previous test somewhere? They need to be in the `test_controller.json` file. 

Normally the steps would be 

1. `PYTHONPATH=. python3 homekit/init_controller_storage.py -f test_controller.json`
2. `PYTHONPATH=. python3 homekit/pair.py -d ${DEVICEID} -p ${SETUPCODE} -f test_controller.json -a demoTV`
3. Use the pairing e.g. `PYTHONPATH=. python3 homekit/get_accessories.py -f test_controller.json -a demoTV`
4. Unpair the TV: `PYTHONPATH=. python3 homekit/remove_pairing.py -f test_controller.json -a demoTV`

**Comment by Tediore** [comment]
Aaaand I'm dumb. I was running the `python3 homekit/get_accessories.py -f test_controller.json -a demoTV` command literally like that expecting it to work on my end for... some reason. 🤦‍♂ 
I'll run that command with my actual device parameters.

Sorry that I'm so bad at this... I really am trying to help but if you think it's not worth it at this point I wouldn't be offended if you told me that.

**Comment by Tediore** [comment]
Same situation unfortunately: https://paste.ubuntu.com/p/6pHtfgGVjW/
https://paste.ubuntu.com/p/vRKFx8nnyc/

**Comment by jlusiardi** [comment]
can you add this line to `homekit/__init__.py` from the local clone?

```
print("%s: I was imported from %s" %(__name__, __file__))
```

This should print something like `homekit: I was imported from /home/jlusiardi/Dokumente/src/homekit_python/homekit/__init__.py`. If the code from the module does not get loaded from the local clone then it will most likely not print anything.

**Comment by Tediore** [comment]
Ok, just to clear up any potential issues, I uninstalled homekit, deleted everything, recreated the venv, recloned the repo, switched to the include_services... branch and ran the command again. Now there are more visible characteristics:
https://paste.ubuntu.com/p/cr6xXMdsX4/
https://paste.ubuntu.com/p/jNBS5yM8C3/ (JSON)

Again, I'm sorry about all the trouble.

**Comment by jlusiardi** [comment]
No Problem. 

Did you try to read / write some of the characteristics? Like volume up / down on 1.26? 

Perhaps I should implement some more of the Characteristics / Services from https://github.com/KhaosT/HAP-NodeJS/blob/master/src/lib/gen/HomeKit-TV.ts in this repo.
